lib/proxy/service.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"net"
+	"strings"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/lib/reversetunnel"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// proxyService impelements the grpc ProxyService.
+type proxyService struct {
+	clusterDialer ClusterDialer
+	log           logrus.FieldLogger
+}
+
+// DialNode opens a bidrectional stream to the requested node.
+func (s *proxyService) DialNode(stream proto.ProxyService_DialNodeServer) error {
+	frame, err := stream.Recv()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// The first frame is always expected to be a dial request.
+	dial := frame.GetDialRequest()
+	if dial == nil {
+		return trace.BadParameter("invalid dial request: request must not be nil")
+	}
+
+	if dial.Source == nil || dial.Destination == nil {
+		return trace.BadParameter("invalid dial request: source and destinatation must not be nil")
+	}
+
+	log := s.log.WithFields(logrus.Fields{
+		"node": dial.NodeID,
+		"src":  dial.Source.Addr,
+		"dst":  dial.Destination.Addr,
+	})
+	log.Debugf("Dial request from peer.")
+
+	_, clusterName, err := splitServerID(dial.NodeID)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	source := &utils.NetAddr{
+		Addr:        dial.Source.Addr,
+		AddrNetwork: dial.Source.Network,
+	}
+	destination := &utils.NetAddr{
+		Addr:        dial.Destination.Addr,
+		AddrNetwork: dial.Destination.Network,
+	}
+
+	nodeConn, err := s.clusterDialer.Dial(clusterName, reversetunnel.DialParams{
+		From:     source,
+		To:       destination,
+		ServerID: dial.NodeID,
+		ConnType: dial.TunnelType,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
```

Comment:
Looking this over, I'm realizing that until a successful read occurs, the client can't really tell the difference between a failure to dial and an error on an established connection.  I haven't read the client code yet, but I assume that the client code ends up returning a "connection" object upon successful establishment of the stream, rather than upon successful dialing of the target node.

Our ssh-based tunnel dialing typically sends a confirmation to the client after it has successfully dialed, but before starting I/O forwarding, and that is the point at which the client creates its own corresponding connection handle to start I/O with.  I lean toward preferring that model, since it helps to keep the origin of errors a bit clearer.  Maybe we should consider adding a `ConnectionEstablished` frame to be emitted here?

Reply:
Good idea added here [2ebe264](https://github.com/gravitational/teleport/pull/10223/commits/2ebe26457f599ea83458141da0930fd1ee7d8699)